### DISCUSSION TITLE
Colour stack trace using pygments?

### DISCUSSION THREAD

**Discussion by psacawa** [CleanCut/green#219] [Enhancements]
Hello, and thank for this module. 
I am wondering if consideration has been given to adding colour to the stack traces printed out upon failure of a test. For a test runner emphasizing colour, this seems like an obvious omission. I have patched in this functionality using `pygments` in my fork: https://github.com/psacawa/green/commit/1922bc8a877bd41b3195b2469c62d953813cd6e5 in an ad hoc fashion that suits my own use case.
Is there interest in implementing this properly?

**Comment by CleanCut** [comment]
I never considered colorizing tracebacks!  Pygments has no dependencies itself, so nice and tight -- no concerns there.

Would you be interested in trying to polish this up to include in a release?  Here are the items I've noticed that would need to be addressed before merging:

- Restrict colors to the ansi 16-color palette (the ones in the table at the bottom of [this page](https://pygments.org/docs/styles/#terminal-styles)).
  - If it's any help, you can access a `Colors` instance (see `output.py`) via `self.colors` at the place you have made your modification.
  - The reason Green sticks to only these colors is that people will adjust them as part of their terminal settings to make sure that they are readable.  If you choose absolute colors, then they will tend to become unreadable depending on some folks terminal background, but the "standard" 16 colors are nearly always remapped to a readable color according to the user's color scheme.  I'd love to use 256 colors (or more!), but then the project would be unreadable for a lot of people.
- If the user has explicitly disabled colors, you should not attempt to colorize the traceback.  At the point where you are making modifications, you can do `if self.colors.termcolors == False:  (color tracebacks)   else:  (don't color tracebacks)`
- Add tests in green/test/test_result.py for the various forms of output, at the very least:
  - Colorizing a known traceback works like you expect
  - If the user has specified `--notermcolor`, then don't colorize the traceback.
- Add `pygments` as a dependenciy to `setup.py` and `requirements.txt`

**Comment by psacawa** [comment]
I've made a commit that i believe addresses your concerns: https://github.com/psacawa/green/commit/d96cbff67cfe8d5de96558c8907f2bd5a48cf4af
It uses `pygments` `TerminalFormatter`. Browsing its source, it appears to use only the 16 colour identifiers referenced on that page you linked. 
It doesn't support `pygments` styles, so there is no need for a command line arg, unless perhaps to turn off stack trace colouring independent of `--notermcolor`. My thought is that this colouring should be on by default, so I didn't add an arg.
There are no tests right now. If they are added, is the current change able to be pulled?

**Comment by CleanCut** [comment]
@psacawa Ah, I failed to mention to please make sure the pygments output uses one of the existing colors we are already using (Red, Green, Blue, Yellow, Bright White).  Right now the two offenders that I see are `AssertionError` is a Bright Red instead of Red, the strings are Cyan -- let's do Yellow instead.

There is already a `--notermcolor` flag, which you just added support for (it's what sets `self.colors.termcolors`).

Yes, if you add tests and find a way to adjust the colors, this may be ready to merge.
